Update player web interface (v0.7.0)

chme opened the pull request:
- Add settings for artwork sources (closes #908)
- Add pop up dialog for Remote pairing requests
- Support adding / removing podcast subscriptions (pr #929)
- Support marking all new podcast episodes / all episodes of one podcast as played (pr #929)
- Support searching by smart pl queries (pr #931)
- Update dependencies
- Update to Vue CLI 4 (includes a lot of linter changes)
- Small style changes (e. g. update library buttons in "About" page)
- Skip buttons for audiobooks and podcasts (pr #729)
- Generate colored placeholder image, if cover artwork is missing on the now playing page
- Styling changes of the navbars and moving the volume controls from the top to the bottom navbar
---

Settings for artwork sources and Remote pairing screenshots:
<img src="https://user-images.githubusercontent.com/1037620/77000020-eef00500-6957-11ea-9aa6-2a471f66f5bf.png" width="300"> <img src="https://user-images.githubusercontent.com/1037620/76999854-a7697900-6957-11ea-93f6-9dcd7d13db4d.png" width="300">


Podcast screenshots:
<img src="https://user-images.githubusercontent.com/1037620/79063460-1fe6ff80-7ca2-11ea-92d3-54ce628792dc.png" width="300"> <img src="https://user-images.githubusercontent.com/1037620/79063464-2aa19480-7ca2-11ea-9a15-e78f273e6f50.png" width="300"> <img src="https://user-images.githubusercontent.com/1037620/79063473-3a20dd80-7ca2-11ea-913f-75e3704e77ac.png" width="300"> <img src="https://user-images.githubusercontent.com/1037620/79063481-46a53600-7ca2-11ea-96f7-5be016f20ed4.png" width="300"> 

Search with smart pl query:
<img src="https://user-images.githubusercontent.com/1037620/79063639-19a55300-7ca3-11ea-96d0-da9501d57c54.png" width="300"> 

Navbar bottom and top:
<img src="https://user-images.githubusercontent.com/1037620/79628672-49d96f80-8143-11ea-983e-24ac695365b1.png" width="300"> <img src="https://user-images.githubusercontent.com/1037620/79628744-0f240700-8144-11ea-87b2-2a375d62974d.png" width="300"> 


Now playing page (with and without artwork available):

<img src="https://user-images.githubusercontent.com/1037620/79628688-6f667900-8143-11ea-8288-c24875fe154c.png" width="300"> <img src="https://user-images.githubusercontent.com/1037620/79628693-8c02b100-8143-11ea-9184-a0eb4c7010ec.png" width="300"> 

Skip buttons for podcasts and audiobooks:
<img src="https://user-images.githubusercontent.com/1037620/79628704-b2c0e780-8143-11ea-8a2e-6815cdd20d44.png" width="300">

chme commented:
@whatdoineed2do fyi, i changed a bit more than i first expected regarding the podcast management and the smart query search (see screenshots). I also updated the web interface to use Vue CLI 4 by following the [migration guide](https://cli.vuejs.org/migrating-from-v3/), you probably will have to run `npm install` to get the updated dependencies and update the globally installed Vue CLI (see migration guide).

whatdoineed2do commented on web-src/src/components/ModalDialogAlbum.vue:
```diff
@@ -12,8 +12,9 @@
-                <a class="button is-small" @click="mark_played">Mark as played</a>
+              <div class="buttons" v-if="media_kind === 'podcast'">
+                <a class="button is-small" v-if="new_tracks > 0" @click="mark_played">Mark as played</a>
+                <a class="button is-small" @click="$emit('remove_podcast')">Remove podcast</a>
```
would it be possible to add `is-danger` modifier to the `Remove podcast`?

chme replied:
I don't want to make the button that prominently visible. It only has a function for podcasts added over the JSON API. For podcasts from the local library this button only leads to an error message.

To avoid accidental deletion, there is now a confirmation dialog displayed before actually removing the podcast.

whatdoineed2do commented on web-src/src/components/ModalDialogAlbum.vue:
```diff
@@ -12,6 +12,10 @@
                 <a class="has-text-link" @click="open_album">{{ album.name }}</a>
               </p>
+              <div class="buttons" v-if="media_kind === 'podcast'">
+                <a class="button is-small" v-if="new_tracks > 0" @click="mark_played">Mark as played</a>
```
On the `podcasts` page, when this dialog is brought up it does not allow us to _Mark as played_ as it doesn't have a `new_tracks` reference
```
v-if="new_tracks === undefined || new_tracks > 0"
```
Can we have this _Mark as Played_ displayed unconditionally here?  If there's no unplayed tracks theres no updates to the db.  In the indivudla `podcast` page, this button would only exit if its really got new items.

The use case for this:

if I add a number of new RSS podcasts at the same time/(adjust the limit of an a few existing subscriptions)[https://github.com/ejurgensen/forked-daapd/issues/951#issuecomment-615392753], the `podcasts` page may have a long list of unplayed items.  I want to mark played for individual podcasts without going into the `podcast` page to do this.

chme replied:
Yes, makes sense. I'll change that.

whatdoineed2do commented:
On the `ModalDialogAlbum` the album art is at the top - when the art is slow to load (most of the time) it makes the new _Mark as played_ or _Remove podcast_ buttons jump/moved when the image loads which is annoying.

Can you consider moving the artwork to the bottom of that card?

chme commented:
> Can you consider moving the artwork to the bottom of that card?

I'll experiment a bit with that. But my first instinct is, to keep it at the top and instead show a placeholder to avoid the jumping of the contents.

whatdoineed2do commented:
this'll work - if theres no artwork can you just leave a placeholder to avoid the jumpiness of removing the placeholder?

davidlb commented:
The JSONAPI now provides a way to set the number/limit of podcasts to retrieve for RSS (3994ef0f2eef97826363307d8394175b111686b0), it would be nice to have this feature in the web interface too.

whatdoineed2do commented:
Any thoughts of integrating #716 into this web update with the work ongoing with `now playing` revamp?

The other PR covers:
* showing `rating` on `now playing` page as 0-5 stars
* allow users to set the rating on `now playing`
* filtering at `album` page on rated tracks

Having 1 and 2 from the list above would be a good addition

chme commented:
@whatdoineed2do I looked into #716 but i want showing the rating stars to be optional. There should be a new setting (probably exposed in the "Webinterface" section of the settings page). I have some other settings i want to expose in the settings page, and will look into #716 when i am doing that.

ejurgensen commented:
Good to have this merged, a lot of nice new features. I have tried it a bit and had a few observations:

1. On iPhone 8/Safari it is not possible to scroll all the way down to the bottom of the burger menu (it "bounces back"). It makes it difficult to select Settings, and impossible to select About.
2. "Spotify" as artwork source only works if you have logged in - should there be a note about that? An alternative might be to have the option as part of the Spotify options in "Online Services".
3. The way one finds the speakers is perhaps a bit too subtle? I took me a while to realise that the small arrow opened the menu.
4. It's good to have the skip forward/backward for podcasts, but I associate the icons used for that with fast forward/rewind.
5. Times/dates are shown in US format - is it possible to show them localised?

chme commented:
> 1. On iPhone 8/Safari it is not possible to scroll all the way down to the bottom of the burger menu (it "bounces back"). It makes it difficult to select Settings, and impossible to select About.

Mmh ... strange ... looking through the changes, nothing jumps out that could have caused it. Sounds a lot like the report #978 for the bottom bar. Do you experience #987 as well (maybe try in landscape mode to get an overflow of the speaker list)?

> 2. "Spotify" as artwork source only works if you have logged in - should there be a note about that? An alternative might be to have the option as part of the Spotify options in "Online Services".

I fixed the conditional display of the Spotify artwork source (instead of checking if forked-daapd was compiled with Spotify support, it now checks if the user has logged in). I will not be displayed, if the user is not logged in.

> 3. The way one finds the speakers is perhaps a bit too subtle? I took me a while to realise that the small arrow opened the menu.

I get what you mean. Do you have a recommendation? Maybe changing the icon to a volume icon?

> 4. It's good to have the skip forward/backward for podcasts, but I associate the icons used for that with fast forward/rewind.

I plan to make the skip intervals changeable via a setting and then the time in seconds that will be skipped should be overlayed over the icon (like AntennaPod does it). Unfortunately Bulma does not offer this kind of overlays out of the box. There exists Bulma extension that solve it, but for that we will have to use a css preprocessor (sass/scss). I probably will look into this soon(ish).

> 5. Times/dates are shown in US format - is it possible to show them localised?

Yes, they look a bit weird for me too. It's on my todo list (but not with an especially high priority ...). Maybe you can open this as a separate issue, so that it keeps nagging me :-)

ejurgensen commented:
> I get what you mean. Do you have a recommendation? Maybe changing the icon to a volume icon?

Yes, or a speaker icon. Perhaps a combination of an icon with the current arrow?

> I plan to make the skip intervals changeable via a setting and then the time in seconds that will be skipped should be overlayed over the icon

Sounds nice. I like those circular skip ahead icons, e.g. [like this one](https://icons8.com/icon/57935/skip-ahead-15-seconds).

chme commented:
@ejurgensen can you test your point 1 again with the current version of #980 ? And if it still does not work, could you post a screenshot?

ejurgensen commented:
The issue in point 1 is still present. Below is a screenshot where I have scrolled to the bottom.

![scroll](https://user-images.githubusercontent.com/1156686/80868432-d2572480-8c9a-11ea-9663-6df0143742b5.png)

chme commented:
> The issue in point 1 is still present. Below is a screenshot where I have scrolled to the bottom.

Added another change, that hopefully fixes this issue. My guess is, that the visible browser top bar is not taken into account when Safari calculates the visible height and therefor the scrollbar in the menu is missing. I have added a margin to the bottom of the navbar menu, that should workaround this issue. Can you give this another try @ejurgensen ?

ejurgensen commented:
Yes, the issue is gone!
